### DISCUSSION TITLE
Deploy m2e core bundles to maven central

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-m2e/m2e-core#1388] [Ideas]
There are some projects that are related to m2e (e.g. tycho m2e targets, jdt-lsp, ...) and currently these would be forced to use the P2 site, but now days where it is possible to consume maven items in PDE/Tycho directly it seems more reasonable to share the code via maven-central.

I therefore like to suggest that we build a way to deploy some of the core bundles also to maven, this list might be inspired by external consumers, and should for now include:

- org.eclipse.m2e.core
- org.eclipse.m2e.pde.target
- ...
- maybe the runtime as well

**Comment by HannesWell** [comment]
Sounds good.
This would also help for issues like https://github.com/eclipse-m2e/m2e-core/issues/830.

I suggest we target this for the next m2e release that will likely happen in the context of Eclipse SimRel 2023-09?

**Reply by laeubi** [reply]
Actually I don't think we are bound to any specific release-cycle here and can do it whenever desired, we probably need to assign a groupId for m2e first (not sure how the process here is for eclipse projects).

**Reply by mickaelistria** [reply]
IIRC, the groupId is requested to and then attributed by Maven Central admins. I think you can just request `org.eclipse.m2e` and everyone will be OK with it.
Also, I'm curious: if some bundles are deployed to Central, do you envision pushing them all? And then building a p2 repo that just aggregates metadata and references those artifacts from Central? Or still building a "regular" p2 repo with the cloned artifacts and keep using it in p2 land? Not that I prefer any personally, I think they're functionally equivalent, I'm just really curious about the longer-term vision.

**Reply by laeubi** [reply]
Ideally we would deploy all to maven central and the simrel-aggregator can consume it from there, but haven't tested it so far and there might be general concerns.
Neverless we can try a hybrid-strategy like platform does so that consumers can migrate to the maven ones (and desire dependabot update notifications) instead of a P2 site.

**Comment by mickaelistria** [comment]
> Ideally we would deploy all to maven central and the simrel-aggregator can consume it from there

I don't think SimRel aggregator has ability to directly consume from Maven; but more importantly, I don't foresee the Eclipse p2 UI (Help > Install New Software, or Eclipse Markerplace Client) getting capable of directly installing artifacts from Maven coordinates.
So I think a p2 repo will still be necessary for long. Fortunately, Tycho has a plugin that allows to create a p2 repo that mostly reference Maven artifacts from Maven repo (instead of including them), this could already be used, even in a hybrid strategy.

**Reply by laeubi** [reply]
Tycho can generate a `maven-p2-repository` you can use directly in P2, but this requires to ship m2e of course to be used. Nerveless I know that @merks has written a converter that converts the jetty maven p2-repo into a traditional one, so I think we have all tools potentially needed at hand.

**Reply by merks** [reply]
I didn't implement a converter.  I just reused the awesomeness of Tycho (and wrapped it with some publishing infrastructure):
```
      <plugin>
        <groupId>org.eclipse.tycho.extras</groupId>
        <artifactId>tycho-p2-extras-plugin</artifactId>
        <version>${tycho-version}</version>
        <executions>
          <execution>
            <phase>generate-resources</phase>
            <goals>
              <goal>mirror</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <source>
            <repository>
              <id>jetty</id>
              <layout>p2</layout>
              <url>mvn:org.eclipse.jetty:jetty-p2:${jetty-version}:zip:p2site</url>
            </repository>
          </source>
          <compress>true</compress>
          <xzCompress>true</xzCompress>
          <keepNonXzIndexFiles>false</keepNonXzIndexFiles>
          <followStrictOnly>true</followStrictOnly>
        </configuration>
      </plugin>
```
